relations/jf/src/lib.rs

```diff
@@ -26,18 +27,16 @@ pub fn generate_srs<R: CryptoRng + RngCore>(
 }
 
 /// Common API for all relations.
-pub trait Relation: Default {
-    /// Public input to the relation. Must be marshallable.
-    type PublicInput: Marshall;
-    /// Private input to the relation.
-    type PrivateInput;
-
-    /// Constructs new relation object from public and private inputs.
-    fn new(public_input: Self::PublicInput, private_input: Self::PrivateInput) -> Self;
-
-    /// Include this relation in the circuit.
-    fn generate_subcircuit(&self, circuit: &mut PlonkCircuit<CircuitField>) -> PlonkResult<()>;
+pub trait Relation {
+    /// Include this relation in the circuit and return all variables that are needed for outside circuit.
+    fn generate_subcircuit(
+        &self,
+        circuit: &mut PlonkCircuit<CircuitField>,
+    ) -> PlonkResult<Vec<Variable>>;
+}
 
+/// Common API
+pub trait ProofSystem: Relation + Default {
```

Comment:
This sets of my spidey-senses: it feels like there shouldn't exist a default proof system or exist _some_ _single_ `Relation` instance for this to type-check. 

Instead of these trait bounds we could pass the `Relation` as an argument to methods where it's required ( like `generate_keys`).

Reply:
I split in two since the methods from the second trait are not needed for NoteRelation.

Reply:
but they come with default implementation, so no effort needed

Reply:
There is mental effort and poor rustc has to generate that :P I don't insist, which version you prefer @deuszx? Going back to one trait or two traits and the second in version you proposed.

Reply:
outdated